docs: add brew livecheck documentation (#7937)
Homebrew/brew: docs/Brew-Livecheck.md

2. **Research available sources**. It's generally preferable to check for a new version at the same source as the `stable` URL, when possible. With this in mind, it may be a good idea to start by removing the file name from the `stable` URL, to see if this is a directory listing page. If that doesn't work, the website may have a downloads page we can check for versions. If it's not possible to find the newest version at this source through any means, try checking other sources from the formula (e.g. an upstream Git repository or the homepage). It's also sometimes necessary to search for other sources outside of the formula.

3. **Compare available versions between sources**. If the latest version is available from the `stable` source, it's best to use that. Otherwise, check the other sources to identify where the latest version is available.

4. **Select a source**. After researching and comparing sources, decide which one is the best available option and use it as the `url` in the `livecheck` block.

> 5. **Create a regex, if necessary or beneficial**. If the check works fine without a regex and wouldn't benefit from having one, it's fine to omit it. However, when a default check isn't working properly and we need to create a `livecheck` block, a regex is almost always necessary as well. More information on creating regexes can be found in the [regex guidelines](#regex-guidelines) section.

Author: When I integrated all the feedback for this section, it was very short and I felt that it was a lot less useful for people new to livecheck. I've tried to find some middle ground between my original version and the requested feedback.

6. **Verify the check is working as intended**. Run `brew livecheck --debug <formula>` again to ensure livecheck is identifying all the versions it should and properly returning the newest version at the end.

### URL guidelines

* **A `url` is a required part of a `livecheck` block** and can either be a string containing a URL (e.g. `"https://www.example.com/downloads/"`) or a symbol referencing one of the supported formula URLs (i.e. `:stable`, `:homepage`, or `:head`).

* **Use a symbol for a formula URL (i.e. `:homepage`, `:stable`, and `:head`) when appropriate**, to avoid duplicating formula URLs in the livecheckable.

* **It's generally preferable to check for versions in the same location as the stable archive, when possible**. This preference is stronger for first-party sources (websites, repositories, etc.) and becomes weaker for third-party sources (e.g. mirrors, another software package manager, etc.).

### Regex guidelines

> The regex in a `livecheck` block is used within a strategy to restrict matching to only relevant text (or strings) and to establish what part of the matched text is the version (using a capture group).

Reviewer: ```suggestion
The `livecheck` block regex is used to restrict matches to a subset of the text and uses a capture group to match the version.
```

> Creating a good regex is a balance between being too strict (breaking easily) and too loose (matching more than it should).

Reviewer: I think this goes without saying?

> * For technical reasons, **the `regex` call in the `livecheck` block should always use parentheses** (e.g. `regex(/example/)`).

Reviewer: Agreed, a RuboCop one specifically.

> * For technical reasons, **the `regex` call in the `livecheck` block should always use parentheses** (e.g. `regex(/example/)`).

Author: We have a RuboCop that enforces parentheses for the `regex` call these days (thanks to Nanda), so I've removed this guideline.

> * **Regex literals should follow the established Homebrew style**, where the `/.../` syntax is the default and the `%r{...}` syntax is used when a forward slash (`/`) is present.

Reviewer: ```suggestion
```

This is already enforced by `brew style`/`rubocop`/

> * **Regexes should adequately represent their intention.** This requires an understanding of basic regex syntax, so we avoid issues like using `.` (match any character) instead of `\.` when we only want to match a period. We also try to be careful about our use of generic catch-alls like  `.*` or `.+`, as it's often better to use something non-greedy and contextually appropriate. For example, if we wanted to match a variety of characters while trying to stay within the bounds of an HTML attribute, we could use something like `[^"' >]*?`.

Reviewer: ```suggestion
* Avoid generic catch-alls like  `.*` or `.+`. Use something non-greedy and contextually appropriate e.g. to match characters within the bounds of an HTML attribute use `[^"' >]*`.
```

`[^"' >]*?` (doesn't really make sense because the `*` is any number of matches anyway)

Author: When `?` comes after `*` or `+`, it makes it non-greedy. That is to say, `.*` and `.*?` have different behaviors.

In practice, a greedy match can escape the bounds of an HTML attribute, matching everything until the last match on the page. We want to ensure we only match text within the current attribute, so it's important to make this non-greedy (i.e., `[^"' >]*?` or `[^"' >]+?`).

This is my understanding from what I've seen in practice but please correct me if I'm wrong.

> * **Try not to be too specific in some parts of the regex.** For example, if a file name uses a hyphen (`-`) between the software name and version (e.g. `example-1.2.3.tar.gz`), we may want to use something like `example[._-]v?(\d+(?:\.\d+)+)\.t` instead. This would allow the regex to continue matching if the upstream file name format changes to `example.1.2.3.tar.gz` or `example_1.2.3.tar.gz`.

Reviewer: ```suggestion
* When a file name uses a hyphen (`-`) between the name and version (e.g. `example-1.2.3.tar.gz`) use `example[._-]v?(\d+(?:\.\d+)+)\.t` instead. This will continue matching if the file name format changes to `example.1.2.3.tar.gz` or `example_1.2.3.tar.gz`.
```

But seems like this should be a RuboCop check.

Author: We may not be able to create a RuboCop for this guideline (yet). It seems like we would need to be able to identify what part of the regex is the software name and what part is the version to say, "Use `[._-]` here instead of _____".

I'll tinker with it when I work on adding more RuboCops but I'm not sure we can make something that won't erroneously warn about things it shouldn't.

> * **Regexes should be case insensitive unless case sensitivity is explicitly required for matching to work properly**. Case insensitivity is enabled by adding the `i` flag at the end of the regex literal (e.g. `/.../i` or `%r{...}i`). This helps to improve reliability and reduce maintenance, as a case-insensitive regex doesn't need to be manually updated if there are any upstream changes in letter case.

Reviewer: ```suggestion
* Regexes should be case insensitive when possible by adding `i` after the regex literal (e.g. `/.../i` or `%r{...}i`). This improves reliability; it doesn't need to be updated if there are any changes in case.
```

> * **Regexes should only use a capturing group around the part of the matched text that corresponds to the version**. For example, in `/href=.*?example-v?(\d+(?:\.\d+)+)(?:-src)?\.t/i`, we're only using a capturing group around the version part (matching a version like `1.2`, `1.2.3`, etc.) and we're using non-capturing groups elsewhere (e.g. `(?:-src)?`). This allows livecheck to rely on the first capture group being the version string.

Reviewer 2: Why would you need a second capture group?

Reviewer: Alternatively: if the second capture group is unused anyway: why does this matter as long as the first one is the one matching the version?

Author: I've massaged this language a bit to assert that capturing groups should only be used around version text, as there seemed to be some confusion around the "first capture group" language.

> Why would you need a second capture group?

In the overwhelming majority of cases, a regex only contains one capture group: the version text. We now have a small number of regexes that contain more than one capture group, where the regex is used within a `strategy` block to combine the captured text into one version (e.g., a cask version like `1.2.3, abcdefgh`).

> Alternatively: if the second capture group is unused anyway: why does this matter as long as the first one is the one matching the version?

It's one of those situations where it's fine until it's not. It's easier and more reliable to say, "Only use a capture group around the version text" instead of "you can use a capture group as long as it comes after all the capture group(s) for version text". The latter case would allow people to be lazy about this rule in most cases but it would come back to bite us.

For a practical example, let's say an upstream repository historically tags versions like `1.2.3`. The standard regex for Git tags like this (`/^v?(\d+(?:\.\d+)+)$/i`) only has a capture group around the version, so it works fine. Over time, upstream switches to a `release-1.2.3` tag format but they're not always reliable about it, so we end up with `1.2.3`, `release-1.2.4`, `1.2.5`, and `release-1.2.6`.

Since we have to match both of these tag formats, we would need to add an optional prefix at the start of the existing regex. If we're following the strict guideline, we would use `/^(?:release[._-])?v?(\d+(?:\.\d+)+)$/i`, which uses a non-capturing group around the optional `release-` part. This works as expected.

If we aren't strict about this rule, someone may naively use a regex like `/^(release[._-])?v?(\d+(?:\.\d+)+)$/i`, which would treat `release-` as the version. This `release-` version is technically newer than `1.2.5` (due to how `Version` comparison works) and this check would pass CI since livecheck technically found a version. If a maintainer doesn't manually test the `livecheck` block outside of CI (I always do) or check the CI log to catch this, it would be merged and break the check.

---

That said, I believe we have to be strict about this anyway to be able to enforce it using a RuboCop. From an implementation standpoint, I imagine this RuboCop will basically be "ensure there's only one capture group in the regex when there isn't a `strategy` block".

> * **Regexes should only match stable versions**. Regexes should be written to avoid prerelease versions like `1.2-alpha1`, `1.2-beta1`, `1.2-rc1`, etc.

Reviewer 2: Is there a regex sample for this?

Reviewer 3: I think most of our existing regexes, except for those Formulae where homebrew-core allows pre-releases (such as`recode` or `vbindiff`) satisfy this criterion.

For example, the regex `/^v?(\d+(?:\.\d+)+)$/i`, which we use when matching git tags avoids matching these pre-release versions. If this example is good enough, it could be added to the doc.

Reviewer: This seems like a poor fit for living in every regex and instead something that `livecheck` itself should know to strip out.

Author: For what it's worth, I removed this guideline from the most recent draft. How to avoid matching unstable versions depends on the context of a given formula/cask, so it can be hard to explain and may not be something that we can enforce through automation anyway.

---

> This seems like a poor fit for living in every regex and instead something that livecheck itself should know to strip out.

We have some simple logic in `Livecheck#latest_version` that's intended to filter out unstable versions ([logic here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/livecheck.rb#L573-L580) and the [`UNSTABLE_VERSION_KEYWORDS` array here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/livecheck.rb#L38-L47)). However, this currently only applies when the formula/cask doesn't have a `livecheck` block. Some formulae/casks use an unstable version, so we have to be able to match unstable versions using a `livecheck` block.

In general, `Version` has some built-in handling of alpha/beta/pre/rc, so it seems like this should probably be something that `Version` detects and `livecheck` simply checks. We're not allowed to access the underlying `tokens` of a `Version` (see #7340), so this would require something like `Version#stable?` (and/or `Version#unstable?`), which would check if a version's tokens are one of these unstable types.

In general, I think we can improve the unstable version filtering logic but I don't think we can reasonably avoid the situation where regexes only match stable versions. Though I put forth the idea of only matching stable versions as a guideline, it's honestly just a side effect of the way we have to create regexes for other reasons.

---

When it comes to Git tags, we create a regex that only matches the desired format because we frequently have to filter out all sorts of unexpected/unrelated tags that are perpetually treated as newest (from the standpoint of `Version` comparison). These tags prevent the check from working properly and it's necessary to exclude them. To illustrate why we can't use a looser regex, I'll provide some examples.

Say we have a repository with a stable tag format like `1.2.3`, unstable tags like `1.2.3-rc4`, and a weird tag like `1.2.3-some-fix`. If we use a regex that allows it to match unstable versions (i.e., not restricting the end), it will also match `1.2.3-some-fix` and this will be treated as newest until a later version shows up (e.g., `1.2.4`).

As another example, a number of repositories contain tags for several different pieces of software. Some will use the `1.2.3` tags for the main software in the repository and use a prefix for other versions, like `something-5.6.7`.

If we're loose about the start of the regex, then it would match tags like `1.2.3` as well as `something-5.6.7`. Since we only capture the version part, it will return a version that looks plausible (`5.6.7`) but is actually related to a different piece of software that's not used in the formula/cask. This particular type of problem is only addressed when someone says, "Hm, that doesn't look right", checks upstream, and corrects the issue using a `livecheck` block.

Basically, there are benefits to restricting the format of Git tags and this naturally filters out unstable versions in the process.

---

It's kind of a similar situation with file name regexes. If we have a regex like `/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.t/i`, the trailing `\.t` is primarily intended to only match versions from tarballs. However, a side effect is that it implicitly excludes unstable versions, as it would match `example-1.2.3.tar.gz` but not `example-1.2.4-rc1.tar.gz`.

To match unstable versions, we would need to add something like `[^"' >]*?` at the end of the capturing group, like `(\d+(?:\.\d+)+[^"' >]*?)`. This would match `example-1.2.4-rc1.tar.gz` and produce a version like `1.2.4-rc1`.

It would need to be done this way because the unstable indicator (alpha/beta/rc/pre) must be part of the version text for livecheck to [theoretically] filter it out without using the regex. If we allowed a looser match outside of the capture group (i.e., `/href=.*?example[._-]v?(\d+(?:\.\d+)+)[^"' >]*?\.t/i`), this would also match unstable versions but the returned version would appear to be stable (i.e., `1.2.3-rc4` would appear as `1.2.3`).

As before, this would match more than just unstable versions. If we encounter a weird version that we don't want but it isn't filtered out as an unstable version, then it could erroneously appear as newest to livecheck and essentially break the check.

Since we would have to actively loosen file name regexes to match unstable versions and this would open us up to potential issues, it isn't something that I've seen value in pursuing. To me, it would be more work for a worse result without a compelling reason behind it (i.e., we're generally only interested in stable versions anyway).

There are a small number of formulae/casks using an unstable version and we do actively expand those regexes to be able to match unstable versions. However, if/when the formula/cask returns to a stable version, we would return the regex to a normal format.

---

Does this help to explain things?

> * **Restrict matching to `href` attributes when targeting file names in an HTML page (or `url` attributes in an RSS feed)**. Using `href=.*?` (or `url=.*?`) at the start of the regex will take care of any opening delimiter for the attribute (`"`, `'`, or nothing) as well as any leading part of the URL. This helps to keep the regex from being overly specific, reducing the need for maintenance in the future. A regex like `href=.*?example-...` is often fine but sometimes it's necessary to have something explicit before the file name to limit matching to only what's appropriate (e.g. `href=.*?/example-...` or `href=["']?example-...`). Similarly, `["' >]` can be used to target the end of the attribute, when needed.

Reviewer 2: I'm not sure what homebrew wants from me here. Should I use href? Should I not? What do I need to use it with?

Can we work with some examples here?

Reviewer: Having read this I have no idea when I should or shouldn't use this.

Author: Basically, if we're matching a file name on an HTML page, restricting matching to `href` attribute URLs will reduce the scope and help to avoid unwanted matches. I reworked this particular guideline to the following:

* **Anchor the start/end of the regex, to restrict the scope**. For example, on HTML pages we often match file names or version directories in `href` attribute URLs (e.g. `/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.zip/i`). The general idea is that limiting scope will help exclude unwanted matches.

Some of the previous information is handled by related examples in the new "Example `livecheck` blocks" section, which hopefully does a better job of providing context. Let me know if these areas are still unclear and I'll try to clarify further.

Reviewer 4: @samford This is tangential, but as I’m seeing `(\d+(?:\.\d+)*)` a lot, why is such a complex regex needed? Specifically, which case does `(\d+(?:\.\d+)*)` correctly catch that `([\d\.]+)` wouldn’t?

Author: The reason we use `v?(\d+(?:\.\d+)+)` (note the `+` instead of `*`) is because it will ensure we're only matching something like `1.2`/`v1.2`, `1.2.3`/`v1.2.3`, etc. When I started working on improving livecheck over a year ago, `[0-9.]+` was used to match version text and the looseness was sometimes breaking checks due to it matching unwanted text and returning it as the newest version.

We created the more explicit regex in response to those issues and have standardized on using it when we want to match a version like `1.2`/`v1.2.3`, etc. People see the relative complexity of this regex and suggest something simpler like `[\d.]+` but we have experience with that not being sufficient.

The only problem we've had with the `+` regex is that it sometimes needs to be loosened to `v?(\d+(?:\.\d+)*)` (i.e., `*` instead of `+`) but this doesn't happen very often. For example, we use the `*` regex if a project predictably publishes the first release for a major version as `2` instead of `2.0`/`2.0.0` (e.g., `openjdk`). We also sometimes use the `*` regex for date-based versions that don't use delimiters (e.g., YYYYMMDD) but we're often more explicit with the regex when a date-based version does use a delimiter (e.g., `(\d{2,4}-\d{2}-\d{2})`).

Despite the looseness of the `*` regex, it's still better than `[0-9.]+` as it won't match text that isn't a complete version like `.`, `.1`, etc. It's explicit enough that it will only match text like `1`, `123`, `1.0`, etc. and this helps to avoid unexpected situations.

The `+` version is what we use by default (when we won't encounter versions like `1`, `123`, etc.) as the looser `*` version can match something it shouldn't if the text is only composed of digits. For example, a project may use versions like `1.2.3` for stable releases but also publish snapshots using date-based versions like `20210113`. If they both use the same format and are found in the same place, then the looser regex would match both and the date-based version would be erroneously returned as newest. This is something we've encountered a number of times, so this is why we use the `+` regex instead of `*` by default.

To put it into perspective, there are only 45 instances of the `*` regex in homebrew/core, compared to 1000+ instances of the `+` regex.

---

However, I imagine you're primarily seeing `(\d+(?:\.\d+)*)` in homebrew/cask, where there are 330 instances of it (and growing fast). homebrew/cask only has 24 instances of the `+` regex (all created by me, it seems).

Support for casks in livecheck was only recently added and, unfortunately, homebrew/cask has standardized on the wrong regex. I explained this concept in PR review (i.e., use the `+` regex unless `*` is necessary) but it was ignored, unfortunately. I also don't have write access in homebrew/cask, so I don't regularly review `livecheck` block PRs over there (as I can't merge them) like I do in homebrew/core.

Rather than trying to keep up with the pace that `livecheck` blocks are being added to casks, I'm currently trying to focus on adding more RuboCops that will automatically enforce the most of the guidelines. We have some livecheck RuboCops for formulae but they don't apply to casks yet, so I need to figure out how to make them work for both.

---

Regarding the complexity of this regex compared to `[0-9.]+` (or `[\d.]+`), these regexes will be replaced by constants, so the complexity will be abstracted away. At that point, we'll get the benefit of the additional specificity (over something loose like `[0-9.]+`) without asking people to engage with the additional complexity. [I imagine this may also help to improve the current situation with cask regexes not always following guidelines.]

I'm aiming to create a PR that adds these constants soon (the work's pretty much done) but I'm working to finish/merge some other PRs first.

I'm planning to eventually spend time bringing the `livecheck` blocks in homebrew/cask up to standard but I don't plan to work on this until sometime after I've implemented regex constants (as that will require modifying regexes across the entire tap anyway).

* **Use `\.t` in place of `\.tgz`, `\.tar\.gz`, etc.** There are a number of different file extensions for tarballs (e.g. `.tar.bz2`, `tbz2`, `.tar.gz`, `.tgz`, `.tar.xz`, `.txz`, etc.) and the upstream source may switch from one compression format to another over time. `\.t` avoids this issue by matching current and future formats starting with `t`. Outside of tarballs, it's fine to use full file extensions in the regex like `\.zip`, `\.jar`, etc.

> * **When matching versions like `1.2`, `1.2.3`, `v1.2`, etc., use the standard snippet for this in the regex: `v?(\d+(?:\.\d+)+)`**. This is often copy-pasted into the regex but it can also be modified to suit the circumstances. For example, if the version uses underscores instead, the standard regex could be modified to something like `v?(\d+(?:[._]\d+)+)`. [The general idea behind this standard snippet is that it better represents our intention compared to older, looser snippets that we now avoid (e.g. `[0-9.]+`).]

Reviewer: ```suggestion
* This is often copy-pasted into the regex
```

Can we avoid doing this? It's pretty unreadable compared to `[0-9.]+` and we could supply it as a constant instead. This comment applies to everywhere you've referred to "a standard regex" including Git tags, href matching, attribute matching, etc.

Author: As mentioned below, I've planned to implement constants for snippets like these (parts of regexes and entire regexes) for a quite a while and I agree it would likely be a better experience for both contributors and maintainers.

I have a local implementation pretty much done, so I'll wrap it up and create a PR after this docs PR is finished.

* **Similarly, when matching Git tags with a version like `1.2`, `1.2.3`, `v1.2.3`, etc., start with the standard regex for this (`/^v?(\d+(?:\.\d+)+)$/i`) and modify it as needed**. Sometimes it's necessary to modify the regex to add a prefix, like `/^example-v?(\d+(?:\.\d+)+)$/i` for an `example-1.2.3` tag format. The general idea here is that Git tags are strings, so we can avoid unrelated software by restricting the start of the string (`^`) and unstable versions by restricting the end of the string (`$`).

## Built-in strategies

> Livecheck's strategies are established methods for finding versions at either a specific source or a general type of source. The available strategies are as follows:

Reviewer: I'd suggest removing this as it'll go out of date when `livecheck` itself changes and it's not really necessary.

* `Apache`
* `Bitbucket`
* `Git`
* `Gnome`
* `Gnu`
* `Hackage`
* `Launchpad`
* `Npm`
* `PageMatch`
* `Pypi`
* `Sourceforge`

> Each strategy has a `#match?(url)` method which determines whether the strategy can be applied to the provided URL. The `PageMatch` strategy is used as a fallback when a regex is provided and no other strategies apply. `PageMatch` simply uses the regex to match content on a page, so it's the desired strategy for URLs where a more-specific strategy doesn't apply.

Reviewer: Do you need to know about a `match?` method?

> Some of the strategies generate a URL and regex internally. In these cases, the strategy often derives information from the provided URL and uses it to create the URL it will check and the regex used for matching. However, if a `regex` is provided in the `livecheck` block, it will be used instead of any generated regex.

Reviewer: Again: I'm not sure I understand why I'd need t know this.

> Livecheck also has a simple numeric priority system, where 5 is the default unless a strategy has defined its own `PRIORITY` constant. Currently, the `Git` strategy has a higher priority (8) and the `PageMatch` strategy has a low priority (0). In practice, this means that when more than one strategy applies to a URL (usually a specific strategy and `PageMatch`), the higher priority strategy is the one that's used.

Reviewer 2: Maybe just put them in order above and mention that?

Reviewer: Why would I need to know this?

Author: I removed this information as I don't think it's useful in this context.

At the time of the first draft, this information wasn't documented elsewhere. In the interim time, I expanded livecheck's technical documentation, so most of this information can be found on https://rubydoc.brew.sh and I think that's a more appropriate context.

### Tap strategies

> Taps can add strategies to apply to their formulae by creating a `livecheck_strategy` folder in the root directory and placing strategy files within. At a minimum, strategies must provide a `#match?(url)` method and a `#find_versions(url, regex)` method.

Reviewer: What taps already use this? I'm unconvinced this is a better fit than duplication in formulae.

Author: To my knowledge, the only tap that has taken advantage of this was Linuxbrew/homebrew-xorg. There were a number of formulae in that tap checking the same sources and Maxim asked me about allowing taps to provide their own tap-specific strategies. It didn't end up being too difficult to support, so I added it and Maxim created the `Xorg` strategy in homebrew-xorg to handle related formulae without needing to create many individual `livecheck` blocks.

The strategy lived there until some X.org formulae started being integrated into homebrew-core. At that point, I added the `Xorg` strategy to Homebrew/homebrew-livecheck and refactored it. After it was merged, Maxim removed the original strategy from homebrew-xorg. At this point, I'm not aware of any taps with their own strategies but I also haven't looked. It's rare for a tap to use `livecheck` blocks in the first place, so I would be surprised if anyone who's not associated with Homebrew is using a tap strategy.

---

That said, the benefit of allowing taps to create a strategy is that you can potentially avoid having to create `livecheck` blocks for formulae. A strategy will automatically apply to matching URLs, so you can create a strategy that handles groups of formulae without any extra work.

This isn't useful for taps that offer first-party software (where you handle releases yourself and don't need livecheck) but it can be useful for taps that contain third-party software. If enough of the software comes from a specific source, a strategy may make more sense than individual livecheck blocks (especially if they're all very similar).

We only include livecheck strategies in Homebrew/brew that apply to first-party taps, so if someone comes to us and says, "Here's a PR for a strategy that's really useful for my tap" we would probably say, "Sorry, this doesn't apply to any of our formulae/casks and we're not going to include this just for your tap." A response I've seen regarding formulae in homebrew-core is, "Sorry, we're not going to merge this formula but you're free to create a tap for it", so allowing tap strategies is basically in the same vein.

---

That said, I removed this part of the documentation, as it may make more sense to include it in the rubydoc.brew.sh documentation. For the moment, I'm simply leaving this feature undocumented. Folks can discover it by digging into the code and I'll gladly explain it to anyone who comes asking for it but I'm not sure that anyone outside of Homebrew will ever request it.

However, it takes very little code to continue supporting this feature, so I don't think there's any value in dropping it. It may be that some large, external tap (e.g., brewsci/homebrew-bio) could benefit from this in the future.

The `#match?(url)` method takes a URL string and returns `true` or `false` to indicate whether the strategy can be applied to the URL.

`#find_versions(url, regex)` takes a URL and an optional regex and returns a `Hash` with a format like `{ :matches => {}, :regex => regex, :url => url }`. The `:matches` `Hash` uses version strings as the keys (e.g. `"1.2.3"`) and `Version` objects as the values. `:regex` is either the strategy-generated regex (if applicable), the regex provided as an argument, or `nil`. The `:url` is either the strategy-generated URL (if applicable) or the original URL provided.

> The built-in strategies in Homebrew's `livecheck_strategy` folder may serve as examples to follow when creating tap strategies. Many of the built-in strategies simply generate a URL and regex before using the `PageMatch` strategy to do the heavy lifting (e.g. `PageMatch.find_versions(page_url, regex)`). When a strategy is checking a text page of some sort (e.g. HTML, RSS, etc.), it may be able to do the same thing. If a strategy needs to do something more complex, the `Git` and `PageMatch` strategies can be referenced as standalone examples.

Reviewer 2: Existing strategies should be a link.
